packages/babel-parser/src/parser/expression.js

```diff
@@ -492,8 +493,18 @@ export default class ExpressionParser extends LValParser {
   // Parse unary operators, both prefix and postfix.
   // https://tc39.es/ecma262/#prod-UnaryExpression
   parseMaybeUnary(refExpressionErrors: ?ExpressionErrors): N.Expression {
-    if (this.isContextual("await") && this.isAwaitAllowed()) {
-      return this.parseAwait();
+    if (this.isContextual("await")) {
+      if (this.isAwaitAllowed()) {
+        return this.parseAwait();
+      } else {
+        if (this.scope.inFunction) {
+          const lookahead = this.lookahead();
+          if (lookahead.type.startsExpr && !this.isAmbiguousAwait(lookahead)) {
```

Comment:
We can probably avoid the lookahead by moving this logic to the end of this function: after calling `parseUpdate()`, it it returns an identifier whose name is `await`, we can check if it could be an await expression and the call `parseAwait()`.

`parseAwait()` then needs to be modified to accept the `startLoc` and `startPos` parameters (we can get `.start` and `.loc.start` from `parseUpdate`'s result), and call `startNodeAt` instead of `startNode` in `parseAwait`.

Reply:
e16250c